### PR TITLE
[RDY]Batteries as magazines mod

### PR DESCRIPTION
**Pull request by Coolthulhu**
I'm working on turning batteries into more realistic "magazines of power" instead of the current pool of electricity in itemized form.
It will be a battery mod. Simpler to craft than UPS mod, since UPS is mostly sci-fi stuff, while connecting a real car battery to a power converter to power a flashlight/soldering iron/small induction forge should be doable IRL.
Later on, if I manage to get it non-annoying, non-buggy, non-hacky and most importantly convenient (not having to use vehicle menu), regular battery system could be scrapped in favor of the magazine battery one.

So far I managed it to pass the following test:
- Spawn a portable generator
- Pull a storage battery out of it
- Spawn a welder and battery mod
- Mod the welder with the battery mod
- Reload the welder with the storage battery pulled out of the generator
- Weld a sheet iron from some scrap
- Check welder to see that it lost 20 charges, as expected

TODO:
- [X] Add a wrapper for charges check, so that items that check charges (notably soldering iron and mp3 player) work with the magazine battery mod. EDIT: Already exists, it's called `ammo_remaining`.
- [x] Hack around crafting system requiring components to be empty (magazine batteries contain "electricity"). Possibly allow inheriting charges from the used battery?
- [x] Make the system less hacky and hardcoded, so that it can be reused for plutonium batteries (optional)

### CODE REVIEW

**Comment by Coolthulhu** [comment]
Limitations:
* For now it doesn't allow NO_UNLOAD items, most notably the UPS. Excused as "the item has a non-standard battery compartment". Actual reason is that there is no good way of overriding the NO_UNLOAD other than finding all use cases and making them "NO_UNLOAD and also lack of an overriding flag". Not hard, just ugly.
* Items with storage batteries always get bulkier, even if the battery could fit in the original compartment. Could be explained as the mod being reversible and thus not getting rid of the original compartment.
* I might have missed some tools that still check for `charges` directly rather than using `ammo_remaining`

I changed some direct `charges` to `ammo_remaining`, possibly fixing a bunch of tiny bugs in the process. Ditto for `charges` decrements and `ammo_consume`.

I think it's ready for testing. Leaving the [CR] on because some of this may be up to moderately hacky and thus hard to maintain.

Most of the credit goes to @mugling who wrote the magazine code and updated the oxytorch to use it. This PR is mostly just allowing item mods to change non-gun magazines.

**Comment by Coolthulhu** [comment]
Notes:
* This may allow changing balancing of crafting. Currently vehicle electric forges are cheap to use ("realistically" they shouldn't be), but hard to craft tools for. By allowing them to use truck batteries, we could easily have crafting costs of 6000 charges or so.
* Possible future expansion: allowing UPS mods to inherit charges from the battery used to craft them.
* At the moment there is a strict limitation of only one magazine per gun. So no UPS-less coilguns, unfortunately.
* Noisemaker with a truck battery and a field of traps around it may be a bit too good
* Vibrator with a truck battery is a hilarious mental image

**Comment by pisskop** [comment]
bah.

No more broken than putting a flamethrower onto a shopping cart and a small tank and pushing a cargo carrier full of loot and gasoline and flamethrower.

File: data/json/items/vehicle_parts.json [hunk 1]
```diff
@@ -399,48 +405,60 @@
...
         "name_plural" : "car batteries",
         "description" : "A 12v lead-acid battery used to power car electrical systems.",
         "weight" : 18000,
         "to_hit" : -2,
         "color" : "light_cyan",
+        "ammo_type": "electricity",
         "symbol" : ":",
         "material" : ["plastic"],
+        "capacity": 4500,
+        "reliability": 10,
```
Comment (mugling) [inline data/json/items/vehicle_parts.json]: This should be the default in `itype.h` as the current default of `0` is useless in all cases

File: data/json/items/vehicle_parts.json [hunk 2]
```diff
@@ -399,48 +405,60 @@
...
+        "flags": [ "NO_SALVAGE", "NO_UNLOAD", "NO_RELOAD" ],
         "volume" : 3,
         "bashing" : 2,
         "category" : "veh_parts",
         "price" : 2000
     },
     {
-        "type":"GENERIC",
+        "type":"MAGAZINE",
```
Comment (mugling) [inline data/json/items/vehicle_parts.json]: Consider migrating to `vehicleparts` subdir

File: src/activity_handlers.cpp [hunk 3]
```diff
@@ -811,10 +811,10 @@ void activity_handlers::game_do_turn( player_activity *act, player *p )
 
     //Deduct 1 battery charge for every minute spent playing
     if( calendar::once_every(MINUTES(1)) ) {
-        game_item.charges--;
+        game_item.ammo_consume( 1, p->pos() );
```
Comment (mugling) [inline src/activity_handlers.cpp]: Please factor these changes out a a separate PR as they are non-controversial and readily testable and can therefore be merged immediately and reduce the overall size of your PR

File: src/iuse.cpp [hunk 4]
```diff
@@ -203,6 +203,23 @@ void remove_ups_mod( item &it, player &p )
     it.item_tags.erase( "NO_RELOAD" );
 }
 
+void remove_battery_magazine_mod( item &it, player &p )
+{
+    if( !it.has_flag( "BATTERY_AMMO" ) ) {
+        return;
+    }
+    p.add_msg_if_player( _( "You remove the battery compartment mod from your %s!" ), it.tname().c_str() );
+    item mod( "magazine_battery_mod", calendar::turn );
```
Comment (mugling) [inline src/iuse.cpp]: You don't need `calendar::turn` there

File: src/iuse.cpp [hunk 5]
```diff
@@ -2189,124 +2206,139 @@ void remove_battery_mods( item &modded, player &p )
     remove_ups_mod( modded, p );
     remove_double_ammo_mod( modded, p );
     remove_double_plut_mod( modded, p );
+    remove_battery_magazine_mod( modded, p );
 }
 
 int iuse::extra_battery(player *p, item *, bool, const tripoint& )
```
Comment (mugling) [inline src/iuse.cpp]: This is going to get out of hand quickly - I'll elaborate at the end

File: src/iuse.cpp [hunk 6]
```diff
@@ -2600,7 +2635,7 @@ int iuse::extinguisher(player *p, item *it, bool, const tripoint& )
 
 int iuse::rm13armor_off(player *p, item *it, bool, const tripoint& )
 {
-    if (it->charges < it->type->charges_to_use()) {
+    if (it->ammo_remaining() < it->type->charges_to_use()) {
```
Comment (mugling) [inline src/iuse.cpp]: Could use `ammo_required()` or even better you could yank the `ammo_sufficient()` from my recent NPC PR that your currently testing

File: src/iuse.cpp [hunk 7]
```diff
@@ -8050,7 +8085,7 @@ int iuse::cable_attach(player *p, item *it, bool, const tripoint& )
 
 int iuse::shavekit(player *p, item *it, bool, const tripoint&)
 {
-    if (it->charges < it->type->charges_to_use()) {
+    if (it->ammo_remaining() < it->type->charges_to_use()) {
```
Comment (mugling) [inline src/iuse.cpp]: `ammo_required()` or `ammo_sufficient()`

**Comment by mugling** [comment]
> Most of the credit goes to @mugling who wrote the magazine code and updated the oxytorch to use it.

Yes the oxytorch was the demo item but I had always hoped we could have proper batteries so I'm glad you are starting work on an implementation of this

> This PR is mostly just allowing item mods to change non-gun magazines.

This part I'm not so sure about. We have plutonium, double battery, ups and now rechargeable battery mods all with different code paths. If not careful we could end up with toolmod code to mirror the original gunmod implementation.

> Later on, if I manage to get it non-annoying, non-buggy, non-hacky and most importantly convenient (not having to use vehicle menu), regular battery system could be scrapped in favor of the magazine battery one.

The interface is the limiting factor and the reason I stalled on an implementation of this. I think adding tool mods is a misstep and prefer @kevingranade's suggestion of using the standard reloading code with automation (#14797). The interface problem need solving and indeed that should probably be the first step.

> At the moment there is a strict limitation of only one magazine per gun

This I'm going to work on and is one of the reasons I'd prefer all new code to use `ammo_sufficient()` as that abstracts away multiple magazines/ammo to become a decision for `item` and not the consumer.

**Comment by kevingranade** [comment]
> Noisemaker with a truck battery and a field of
> traps around it may be a bit too good

If this becomes a problem it's great, because this should definitely work.
The fix is traps that break over time, not nerfing noisemakers.

**Comment by Coolthulhu** [comment]
I reset the branch to the new, heavily changed master and manually copied the changes.

What needs to be done to the battery mod code for it to be mergeable? The number of hacks is quite low now, just 2 magazine functions.

File: src/item.cpp [hunk 8]
```diff
@@ -4096,9 +4105,19 @@ bool item::magazine_integral() const
 {
     // finds first ammo type which specifies at least one magazine
     const auto& mags = type->magazines;
-    return std::none_of( mags.begin(), mags.end(), []( const std::pair<ammotype, const std::set<itype_id>>& e ) {
+    if( !std::none_of(
```
Comment (mugling) [inline src/item.cpp]: `std::any_of`

File: src/item.cpp [hunk 9]
```diff
@@ -4096,9 +4105,19 @@ bool item::magazine_integral() const
...
+        mags.begin(), mags.end(), []( const std::pair<ammotype, const std::set<itype_id>>& e ) {
         return e.second.size();
-    } );
+    }) ) {
```
Comment (mugling) [inline src/item.cpp]: `!e.second.empty()`

**Comment by Coolthulhu** [comment]
I'm still not sure how should I format the `if(any_of)` block. Is there a standard for that?

**Comment by BevapDin** [comment]
Well, it looks fine and works fine.

Only issues so far: one can not remove the mod as moded items can not be selected when activating a screwdriver. But installing a different mod removes it. You probably have to add the mods tag to the list in `iuse::remove_all_mods`.

**Comment by Coolthulhu** [comment]
Also added the radio mod to `remove_all_mods`, since it removes it, just didn't offer it.

**Comment by Coolthulhu** [comment]
Oops, automerge was weird. Fixed manually, should work now.

**Comment by BorkBorkGoesTheCode** [comment]
Are you abandoning this?

**Comment by Coolthulhu** [comment]
It was open and mergeable for a while. Did nobody merge it because it got buried under other PRs or because of quality concerns?

If the former, I could update it.

**Comment by BorkBorkGoesTheCode** [comment]
Merge conflicts.

**Comment by Coolthulhu** [comment]
I mean for roughly the month or so it was free from merge conflicts.

**Comment by Rivet-the-Zombie** [comment]
This looks good. I must have missed it when it was new.

**Comment by BorkBorkGoesTheCode** [comment]
Is ammo_remaining still in the code?

**Comment by BorkBorkGoesTheCode** [comment]
@Coolthulhu @Rivet-the-Zombie Is this broken forever?

**Comment by Rivet-the-Zombie** [comment]
Everything appears to be working properly, except you can use the battery compartment mod on itself, which causes it to disappear entirely.

**Comment by mugling** [comment]
Will take a look in a moment

**Comment by mugling** [comment]
So the code for this PR is clean but I'm not sure about the direction. We currently have item batteries and vehicle batteries. Neither are particularly balanced but both need to remain. I think this PR bridges the gap without solving any of the gameplay/design problems.

Thoughts:
* Gun mod system was horrible and cleanup has taken a long time. Lets not implement tool mods without top-down design. We already have a proliferation of battery mods with many outstanding bugs.

* We should address energy balance before we add new mechanics. Item batteries are both far too abundant and energy dense. A single stack (100) is sufficient to supply your cooking requirements for days on the classic cataclysm oatmeal and toilet water diet.

* Pretty much everyone wanted discrete batteries last time we discussed this (#14797 and #14788). The issue was lack of a sane interface/UX that isn't difficult or awkward to use. This PR doesn't make anything worse but it doesn't provide an improvement either.

I'd prefer to try and do at least some balancing of energy requirements first before anything else. If someone solves the third issue the code support is 100% ready to go and has been for some time so I don't anticipate any significant problems there - indeed the vehicle code already uses 'real' batteries (magazines of charge).

**Comment by Coolthulhu** [comment]
> Lets not implement tool mods without top-down design.

You mean new tool mods or a whole system for them?
There are multiple battery mods already. They're handled in rather coherent way.

> We should address energy balance before we add new mechanics.

I could move it into a mod for now. Ideally the vehicle battery and handheld battery balance would be similar.

> The issue was lack of a sane interface/UX that isn't difficult or awkward to use.

The biggest problem with discrete battery UI is partially charged ones. With giant, bulky vehicle ones this issue is sidestepped.

Though honestly the idea of having to:
Install a battery, turn on the charger, wait, remove battery, reload tool
Sounds like a horrible user experience.

Still, could be a mod, since otherwise it will bitrot and something like this is necessary for a discrete battery system.

**Comment by Coolthulhu** [comment]
Moved to mod.

**Comment by mugling** [comment]
If it needs code support and affects a critical feature there's a strong argument for core? I'll take another look

File: data/mods/Discrete_Batteries/items.json [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+    "ammo": "battery",
+    "name": "battery compartment mod",
+    "description": "A battery compartment mod that allows using high-density, rechargeable vehicle batteries in regular tools.",
+    "price": 2000,
+    "material": ["plastic", "aluminum"],
+    "weight": 500,
+    "volume": 1,
+    "category" : "spare_parts",
+    "use_action": "MAGAZINE_BATTERY",
+    "magazines": [ [
```
Comment (mugling) [inline data/mods/Discrete_Batteries/items.json]: Since you wrote this PR there now is `magazine_adaptor` which the beltfeed item uses

**Comment by mugling** [comment]
> Pretty much everyone wanted discrete batteries last time we discussed this (#14797 and #14788)

The only reason to have a mod is for incubation purposes whilst we work on the design. Competing reloading systems is a world of pain. I'm in favor of this PR providing it's a stepping stone to a complete working design with eventual deprecation of the mod. We don't have the resources to maintain multiple systems in the long term.

> Gun mod system was horrible and cleanup has taken a long time. Lets not implement tool mods without top-down design. We already have a proliferation of battery mods with many outstanding bugs.

First thing lets implement and `islot_toolmod` as an optional slot. Then lets use the existing `magazine_adaptor` and `ammo_modifier` functionality. The `islot_toolmod` should share some fields with `islot_gunmod` via a parent abstract class `islot_item_mod`.

> We should address energy balance before we add new mechanics...

If it starts out as a mod that we later mainline balance can be deferred

> The issue was lack of a sane interface/UX that isn't difficult or awkward to use.

Main concern is crafting. We could sidestep this entirely by allowing the user to mark batteries in the inventory as suitable for providing electrical power during crafting. The code would purposely drain the flagged batteries in a sensible order (least charged first). Otherwise the UX would be identical to the current magazine system.

**Comment by Coolthulhu** [comment]
> First thing lets implement and islot_toolmod as an optional slot.

Here?
The only other toolmod which could use it at the moment would be atomic conversion.
Honestly there isn't that much "advanced" code here - mostly just boilerplate and literally just 30 lines of actual "modding of tools".

Unifying toolmod code could be a good idea, but it would probably double the number of lines and increase the number of "important" lines of code by an order of magnitude.
The most cleanliness to be saved here is from cutting down on iuses, which could be done by moving the code to an actor.

**Comment by mugling** [comment]
> Here?

I'm prototyping something now.

> The only other toolmod which could use it at the moment would be atomic conversion.

Well we also have double ammo and I think a rechargeable one as well?

> Unifying toolmod code could be a good idea, but it would probably double the number of lines and increase the number of "important" lines of code by an order of magnitude.

We already have to check for gunmod adaptors so line count would be the same either way.

**Comment by mugling** [comment]
#18735 would allow for a zero-cost implementation of this PR which leaves us just with balancing which could take place in a temporary incubation mod?

**Comment by Coolthulhu** [comment]
Updated for unified toolmods.
New toolmods required a minor fix: `magazine_integral` did not understand items that stop having integral magazine due to mods. Now it does.

**Comment by mugling** [comment]
Given it's just one item with some inoffensive code changes I'd push this to `core` now.

File: src/item.cpp [hunk 11]
```diff
@@ -4182,10 +4182,21 @@ std::set<std::string> item::ammo_effects( bool with_ammo ) const
 
 bool item::magazine_integral() const
 {
-    // finds first ammo type which specifies at least one magazine
+    // Finds the first mod which changes magazines and returns if it adds a list of them (can just unset instead)
+    // If no mod changes them, checks if there is a magazine set for the base item type
+    auto mods = is_gun() ? gunmods() : toolmods();
```
Comment (mugling) [inline src/item.cpp]: For consistency with other code move this inside the loop initializer

File: data/json/items/toolmod.json [hunk 12]
```diff
@@ -53,5 +53,23 @@
...
+  {
+    "id": "magazine_battery_mod",
+    "copy-from": "mod_battery",
+    "type": "TOOLMOD",
+    "category": "spare_parts",
+    "color": "light_green",
+    "name": "battery compartment mod",
+    "description": "A battery compartment mod that allows using high-density, rechargeable vehicle batteries in regular tools.",
+    "acceptable_ammo": [ "battery" ],
+    "magazine_adaptor": [ [
```
Comment (mugling) [inline data/json/items/toolmod.json]: linting is probably going to want this all on one line

File: src/item.cpp (hunk 11 above)
Comment (Coolthulhu) [inline src/item.cpp]: I copied it from `magazine_compatible`

**Comment by mugling** [comment]
Ok, I'm compiling it now.

Whilst we're at it can you move the other toolmod recipes to the same file and add it it to `json_whitelist`. If any of the balance for the other toolomd recipes offends you it could be changed here but that's by no means a requirement.

**Comment by Coolthulhu** [comment]
I can't get the new linter to work.

**Comment by Coolthulhu** [comment]
OK, managed to get it to work

**Comment by mugling** [comment]
Could try http://dev.narc.ro/cataclysm/format.html

**Comment by mugling** [comment]
I've sent your 3 PR's against this branch that fix #18755

**Comment by mugling** [comment]
Reload prompt is a little wield 'car battery with batteries (2500)'

**Comment by Coolthulhu** [comment]
Seems that for some reason `magazine_current` doesn't work for tools.
Or I'm doing something wrong, will poke around.

**Comment by Coolthulhu** [comment]
Got it.
Special cased NO_UNLOAD, because it doesn't really matter what is loaded in NO_UNLOAD items, only how much ammo is it.

File: src/item.cpp [hunk 13]
```diff
@@ -1399,6 +1399,13 @@ std::string item::info( bool showtext, std::vector<iteminfo> &info ) const
         }
 
         if( !magazine_integral() ) {
+            const auto cur_mag = magazine_current();
```
Comment (mugling) [inline src/item.cpp]: Ideally this should be in the same format as for guns, eg. `Magazine: %s`

Comment (Coolthulhu) [inline src/item.cpp]: I went with a different one because magazine sounds weird here.
Say, battery in a radio would be a "magazine".

Comment (Coolthulhu) [inline src/item.cpp]: Also, should I scrap the ammo display? Guns don't have it.

File: src/item.cpp (hunk 11 above)
Comment (mugling) [inline src/item.cpp]: That should probably change too then

**Comment by Coolthulhu** [comment]
OK, changed to same as in guns and also special case on "battery".

File: src/player.cpp [hunk 14]
```diff
@@ -9537,9 +9537,8 @@ item::reload_option player::select_ammo( const item &base, const std::vector<ite
     std::vector<std::string> names;
     std::transform( opts.begin(), opts.end(), std::back_inserter( names ), []( const reload_option& e ) {
         if( e.ammo->is_magazine() && e.ammo->ammo_data() ) {
-            if( e.ammo->has_flag( "NO_UNLOAD" ) ) {
-                // This ammo is not a real object that can be recovered but pseudo-object
-                // For example, "battery" content of car battery
+            if( e.ammo->ammo_data()->get_id() == "battery" ) {
```
Comment (mugling) [inline src/player.cpp]: `ammo_current()`

**Comment by mugling** [comment]
You still need to merge those 3 PR's against this branch. I don't think there's anything else left to do after that

**Comment by Zireael07** [comment]
Is this mod or core?

**Comment by BorkBorkGoesTheCode** [comment]
A little of both.

**Comment by mugling** [comment]
Core

**Comment by Nhatorama** [comment]
It seems that the in-game item, "battery compartment mod" comes from this mod, if it isn't let me know.

I recently attached the "battery compartment mod" to a makeshift welder and loaded it with a medium storage battery. When it was to time to replace the medium storage with a full one I expected the empty one to go in my inventory or on the ground. I didn't get anything back.

Is this intended? If it is please add it to the description.

**Comment by Coolthulhu** [comment]
> I didn't get anything back.

Could be a bug. It is supposed to treat batteries as magazines, not consume them.
